telemetry/data_feature_usage.go

```
@@ -167,3 +176,70 @@ func GetTxnUsageInfo(ctx sessionctx.Context) *TxnUsage {
 func postReportTxnUsage() {
 	initialTxnCommitCounter = metrics.GetTxnCommitCounter()
 }
+
+// CTEUsage records the usages of CTE.
+type CTEUsage struct {
+	NonRecursiveCTEUsed uint64 `json:"nonRecursiveCTEUsed"`
+	RecursiveUsed       uint64 `json:"recursiveUsed"`
+	NonCTEUsed          uint64 `json:"nonCTEUsed"`
+}
+
+// ResetCTEUsage resets CTE usages.
+func ResetCTEUsage() error {
+	me, err := m.TelemetrySQLCTECnt.GetMetricWith(prometheus.Labels{m.LblCTEType: "nonRecurCTE"})
+	if err != nil {
+		return err
+	}
+	me.Set(0)
+	me, err = m.TelemetrySQLCTECnt.GetMetricWith(prometheus.Labels{m.LblCTEType: "recurCTE"})
+	if err != nil {
+		return err
+	}
+	me.Set(0)
+	me, err = m.TelemetrySQLCTECnt.GetMetricWith(prometheus.Labels{m.LblCTEType: "notCTE"})
+	if err != nil {
+		return err
+	}
+	me.Set(0)
+	return nil
+}
+
+// GetCTEUsageInfo gets the CTE usages.
+func GetCTEUsageInfo(ctx sessionctx.Context) (*CTEUsage, error) {
+	usage := &CTEUsage{}
+	me, err := m.TelemetrySQLCTECnt.GetMetricWith(prometheus.Labels{m.LblCTEType: "nonRecurCTE"})
```

I suggest to ignore errors and continue counting others.